File: src/BackJoon/Sort/수 정렬하기2/quick.java
public class quick {
    public static void main(String[] args) {
        int[] arr = { 675, 235, 59, 100, 48, 1, 3, 354, 3, 234, 6557, 878787, 12 };

        quickSort(arr, 0, arr.length - 1);

        printArr(arr);
    }
    
    public static void quickSort(int[] arr, int start, int end) {
        if (start >= end)
            return;

        int l = start + 1;
        int r = end;

        while (l <= r) {
            while (l <= r && arr[l] < arr[start])
                l++;
            while (l <= r && arr[r] >= arr[start])
                r--;

            if (l <= r) {
                swap(arr, l, r);
            }
        }
        swap(arr, r, start);

        quickSort(arr, start, r - 1);
        quickSort(arr, r + 1, end);
    }
    
    public static void swap(int[] arr, int a, int b) {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

    public static void printArr(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}

File: src/BaekJoon/Sort/수 정렬하기2/quick.js
function qsort(arr, l, r) {
    if (l >= r) return;

    let pivot = l;
    let left = pivot + 1;
    let right = r;

    while (left <= right) {
        while (left <= r && arr[left] <= arr[pivot]) {
            left++;
        }
        while (right > l && arr[right] > arr[pivot]) {
            right--;
        }
        console.log(left + " " + arr[left] + " " + right + " " + arr[right]);
        if (left <= right) {
            // console.log("0");
            swap(arr, left, right);
        }
        // console.log(arr[left] + " " + arr[right]);
    }

    swap(arr, pivot, right);

    qsort(arr, l, right - 1);
    qsort(arr, right + 1, r);
}

function swap(arr, l, r) {
    let temp = arr[l];
    arr[l] = arr[r];
    arr[r] = temp;
}

const arr = [675, 235, 59, 100, 48, 1, 3, 3, 234, 6557, 878787, 12];
qsort(arr, 0, arr.length - 1);

console.log(arr);